File: app/src/main/java/com/dani_chuks/andeladeveloper/presentation_models/mappers/Mapper.java
package com.dani_chuks.andeladeveloper.presentation_models.mappers;

import android.util.Log;

import com.dani_chuks.andeladeveloper.presentation_models.FilmModel;
import com.dani_chuks.andeladeveloper.presentation_models.PersonModel;
import com.dani_chuks.andeladeveloper.presentation_models.PlanetModel;
import com.dani_chuks.andeladeveloper.presentation_models.SpecieModel;
import com.dani_chuks.andeladeveloper.presentation_models.StarshipModel;
import com.dani_chuks.andeladeveloper.presentation_models.VehicleModel;
import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class Mapper {

    public static Gson gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .create();

    public static <Output, Input> Output mapModel(Type type, Input input) {
        Output tp;
//        Type type = new TypeToken<Output>(){}.getType();
        try {
            String jsonString = gson.toJson(input);
            tp = gson.fromJson(jsonString, type);
            return tp;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", type.getTypeName() + ":  ", exception);
        }
        return null;
    }

    public static <Output, Input> Output mapModel2(Output output, Input input) {
        Output tp;
        Type type = new TypeToken<Output>() {}.getType();
        try {
            String jsonString = gson.toJson(input);
            tp = gson.fromJson(jsonString, type);
            return tp;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", type.getTypeName() + ":  ", exception);
        }
        return null;
    }


    public static <Input> ArrayList<FilmModel> mapFims(List<Input> inputs) {
        ArrayList<FilmModel> outputList;
        Type listType = new TypeToken<ArrayList<FilmModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", FilmModel.class + ":  ", exception);
        }
        return null;
    }

    public static <Input> ArrayList<PersonModel> mapPeople(List<Input> inputs) {
        ArrayList<PersonModel> outputList;
        Type listType = new TypeToken<ArrayList<PersonModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", PersonModel.class + ":  ", exception);
        }
        return null;
    }

    public static <Input> ArrayList<PlanetModel> mapPlanets(List<Input> inputs) {
        ArrayList<PlanetModel> outputList;
        Type listType = new TypeToken<ArrayList<PlanetModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", PlanetModel.class + ":  ", exception);
        }
        return null;
    }

    public static <Input> ArrayList<SpecieModel> mapSpecies(List<Input> inputs) {
        ArrayList<SpecieModel> outputList;
        Type listType = new TypeToken<ArrayList<SpecieModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", SpecieModel.class + ":  ", exception);
        }
        return null;
    }

    public static <Input> ArrayList<StarshipModel> mapStarships(List<Input> inputs) {
        ArrayList<StarshipModel> outputList;
        Type listType = new TypeToken<ArrayList<StarshipModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", StarshipModel.class + ":  ", exception);
        }
        return null;
    }

    public static <Input> ArrayList<VehicleModel> mapVehicles(List<Input> inputs) {
        ArrayList<VehicleModel> outputList;
        Type listType = new TypeToken<ArrayList<VehicleModel>>() {}.getType();
        try {
            String json = gson.toJson(inputs, listType);
            outputList = gson.fromJson(json, listType);
            return outputList;
        } catch (JsonSyntaxException exception) {
            Log.d("TAG", VehicleModel.class + ":  ", exception);
        }
        return null;
    }
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/presentation_models/mappers/Mapper.kt
package com.dani_chuks.andeladeveloper.presentation_models.mappers

import android.util.Log
import com.dani_chuks.andeladeveloper.presentation_models.MainModels.*
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import java.lang.reflect.Type
import java.util.*

object Mapper {

//    TODO: Use Reified to refactor line
//     using https://proandroiddev.com/how-reified-type-makes-kotlin-so-much-better-7ae539ed0304

    var gson = GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .create()

    fun <Output, Input> mapModel(type: Type, input: Input): Output? {
        val tp: Output
        //        Type type = new TypeToken<Output>(){}.getType();
        try {
            val jsonString = gson.toJson(input)
            tp = gson.fromJson(jsonString, type)
            return tp
        } catch (exception: Throwable) {
            Log.d("TAG", "$type:  ", exception)
        }

        return null
    }

    fun <Output, Input> mapModel2(output: Output, input: Input): Output? {
        val tp: Output
        val type = object : TypeToken<Output>() {

        }.type
        try {
            val jsonString = gson.toJson(input)
            tp = gson.fromJson(jsonString, type)
            return tp
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", type.toString() + ":  ", exception)
        }

        return null
    }


    fun <Input> mapFims(inputs: List<Input>): ArrayList<FilmModel>? {
        val outputList: ArrayList<FilmModel>
        val listType = object : TypeToken<ArrayList<FilmModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", FilmModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }

    fun <Input> mapPeople(inputs: List<Input>): ArrayList<PersonModel>? {
        val outputList: ArrayList<PersonModel>
        val listType = object : TypeToken<ArrayList<PersonModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", PersonModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }

    fun <Input> mapPlanets(inputs: List<Input>): ArrayList<PlanetModel>? {
        val outputList: ArrayList<PlanetModel>
        val listType = object : TypeToken<ArrayList<PlanetModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", PlanetModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }

    fun <Input> mapSpecies(inputs: List<Input>): ArrayList<SpecieModel>? {
        val outputList: ArrayList<SpecieModel>
        val listType = object : TypeToken<ArrayList<SpecieModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", SpecieModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }

    fun <Input> mapStarships(inputs: List<Input>): ArrayList<StarshipModel>? {
        val outputList: ArrayList<StarshipModel>
        val listType = object : TypeToken<ArrayList<StarshipModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", StarshipModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }

    fun <Input> mapVehicles(inputs: List<Input>): ArrayList<VehicleModel>? {
        val outputList: ArrayList<VehicleModel>
        val listType = object : TypeToken<ArrayList<VehicleModel>>() {

        }.type
        try {
            val json = gson.toJson(inputs, listType)
            outputList = gson.fromJson(json, listType)
            return outputList
        } catch (exception: JsonSyntaxException) {
            Log.d("TAG", VehicleModel::class.java.simpleName + ":  ", exception)
        }

        return null
    }
}
